zcash_primitives/src/note_encryption.rs

```diff
@@ -59,73 +60,127 @@ where
+
 /// An unencrypted memo received alongside a shielded note in a Zcash transaction.
 #[derive(Clone)]
-pub struct Memo([u8; 512]);
+pub enum Memo {
+    /// An empty memo field.
+    Empty,
+    /// A memo field containing a UTF-8 string.
+    Text(TextMemo),
+    /// A future memo format.
+    Future(FutureMemo),
+    /// A memo field containing arbitrary bytes.
+    Arbitrary(Box<[u8; 511]>),
+}
 
 impl fmt::Debug for Memo {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "Memo(")?;
-        match self.to_utf8() {
-            Some(Ok(memo)) => write!(f, "\"{}\"", memo)?,
-            _ => fmt_colon_delimited_hex(f, &self.0[..])?,
+        match self {
+            Memo::Empty => write!(f, "Memo::Empty"),
+            Memo::Text(memo) => write!(f, "Memo::Text(\"{}\")", memo.0),
+            Memo::Future(bytes) => write!(f, "Memo::Future({:0x})", bytes.0[0]),
+            Memo::Arbitrary(bytes) => {
+                write!(f, "Memo::Arbitrary(")?;
+                fmt_colon_delimited_hex(f, &bytes[..])?;
+                write!(f, ")")
+            }
         }
-        write!(f, ")")
     }
 }
 
 impl Default for Memo {
     fn default() -> Self {
-        // Empty memo field indication per ZIP 302
-        let mut memo = [0u8; 512];
-        memo[0] = 0xF6;
-        Memo(memo)
+        Memo::Empty
     }
 }
 
 impl PartialEq for Memo {
     fn eq(&self, rhs: &Memo) -> bool {
-        self.0[..] == rhs.0[..]
+        match (self, rhs) {
+            (Memo::Empty, Memo::Empty) => true,
+            (Memo::Text(a), Memo::Text(b)) => a == b,
+            (Memo::Arbitrary(a), Memo::Arbitrary(b)) => a[..] == b[..],
+            _ => false,
+        }
     }
 }
 
 impl Memo {
-    /// Returns a `Memo` containing the given slice, appending with zero bytes if
-    /// necessary, or `None` if the slice is too long. If the slice is empty,
-    /// `Memo::default` is returned.
-    pub fn from_bytes(memo: &[u8]) -> Option<Memo> {
-        if memo.is_empty() {
-            Some(Memo::default())
-        } else if memo.len() <= 512 {
-            let mut data = [0; 512];
-            data[0..memo.len()].copy_from_slice(memo);
-            Some(Memo(data))
-        } else {
-            // memo is too long
-            None
+    /// Parses a `Memo` from its ZIP 302 serialization.
+    ///
+    /// This API does **not** preserve the exact byte serialization of the memo. For
+    /// example, if the memo is a string containing invalid sequences:
+    /// ```
+    /// use zcash_primitives::note_encryption::Memo;
+    ///
+    /// let text = b"Hello \xF0\x90\x80World";
+    /// let mut bytes = [0; 512];
+    /// bytes[..text.len()].copy_from_slice(text);
+    /// assert_ne!(Memo::from_bytes(bytes).to_bytes()[..], bytes[..]);
+    /// ```
+    pub fn from_bytes(bytes: [u8; 512]) -> Self {
+        match bytes[0] {
+            0xF6 => Memo::Empty,
+            0xF5 | 0xF7 | 0xF8 | 0xF9 | 0xFA | 0xFB | 0xFC | 0xFD | 0xFE => {
```

Comment:
I haven't yet implemented `0xF5` support, because I want to update the ZIP draft first.